lib/stdlib/src/supervisor.erl

```diff
@@ -757,82 +749,72 @@ restart(simple_one_for_one, Child, State0) ->
-	    NChild = lists:keyfind(FailedId, #child.id, NChs),
-	    NChild2 = NChild#child{pid=?restarting(undefined)},
-	    NState = State#state{children = NChs},
-	    {try_again, replace_child(NChild2,NState), FailedId}
+    {Return, NChildren} = restart_multiple_children(Child, Children1, SupName),
+    {Return, State#state{children = NChildren}}.
+
+restart_multiple_children(Child, Children, SupName) ->
+    Children1 = terminate_children(Children, SupName),
+    case start_children(Children1, SupName) of
+	{ok, NChildren} ->
+	    {ok, NChildren};
+	{error, NChildren, {failed_to_start_child, FailedId, _Reason}} ->
+            NewPid = if FailedId =:= Child#child.id ->
+                             restarting(Child#child.pid);
+                        true ->
+                             ?restarting(undefined)
+                     end,
+	    {{try_again, FailedId}, set_pid(NewPid,FailedId,NChildren)}
     end.
 
 restarting(Pid) when is_pid(Pid) -> ?restarting(Pid);
 restarting(RPid) -> RPid.
 
+-spec try_again_restart(child_id() | {'restarting',pid()}) -> 'ok'.
+try_again_restart(TryAgainId) ->
+    gen_server:cast(self(), {try_again_restart, TryAgainId}).
+
 %%-----------------------------------------------------------------
 %% Func: terminate_children/2
-%% Args: Children = [child_rec()] in termination order
+%% Args: Children = children() % ids in termination order
 %%       SupName = {local, atom()} | {global, atom()} | {pid(),Mod}
-%% Returns: NChildren = [child_rec()] in
-%%          startup order (reversed termination order)
+%% Returns: NChildren = children() % ids in startup order
+%%                                 % (reversed termination order)
 %%-----------------------------------------------------------------
 terminate_children(Children, SupName) ->
-    terminate_children(Children, SupName, []).
-
-%% Temporary children should not be restarted and thus should
-%% be skipped when building the list of terminated children, although
-%% we do want them to be shut down as many functions from this module
-%% use this function to just clear everything.
-terminate_children([Child | Children], SupName, Res) when ?is_temporary(Child) ->
-    _ = do_terminate(Child, SupName),
-    terminate_children(Children, SupName, Res);
-terminate_children([Child | Children], SupName, Res) ->
-    NChild = do_terminate(Child, SupName),
-    terminate_children(Children, SupName, [NChild | Res]);
-terminate_children([], _SupName, Res) ->
-    Res.
+    Fun =
+        fun(_Id,Child) when ?is_temporary(Child) ->
+                %% Temporary children should not be restarted and thus should
+                %% be skipped when building the list of terminated children.
+                do_terminate(Child, SupName),
+                remove;
+           (_Id,Child) ->
+                do_terminate(Child, SupName),
+                {update,Child#child{pid=undefined}}
+        end,
+    {ok,NChildren} = ch_map(Fun, Children),
+    NChildren.
```

Comment:
I think it would be easier to understand if you changed Fun to Terminate and ch_map to children_map so that next to last line would read  {ok, Nchildren} = children_map(Terminate, Children).

Reply:
Yes! :)